internal/formatter.go

```diff
@@ -541,6 +541,7 @@ func (n *SubqueryExprNode) FormatSQL(ctx context.Context) (string, error) {
 	}
 	switch n.node.SubqueryType() {
 	case ast.SubqueryTypeScalar:
+		return fmt.Sprintf("(%s)", sql), nil
```

Comment:
I'm not sure about  `case ast.SubqueryTypeLikeAny:` and `case ast.SubqueryTypeLikeAll:`.

But it seems all types of Scalar/Array/Exists/In require parentheses in the end, so it is possible that LikeAny/LikeAll need to be enclosed as well...?

Reply:
Perhaps the parenthesis are needed. If you can add a test case for them, please do.

Reply:
@goccy I added parenthesis to them, but I DID NOT add test cases. [diff](https://github.com/goccy/go-zetasqlite/pull/35/files#diff-0102133126f709542bd93e0b83ea950894fc4ecd89993fe348751b1f8403b38fL561)

:point_down: This is how it happend.

I found that ZetaSQL seems to have extended language feature like `WHERE val LIKE ANY ("abc")` / `WHERE val ALL ANY ("abc")`

https://github.com/google/zetasql/blob/177d495a064e38684c462cf883e22428273bd996/zetasql/public/options.proto#L816
https://github.com/google/zetasql/blob/57651f4f0cf9efa76163bc0282edfcf98dd72414/zetasql/analyzer/testdata/hints.test#L2136

However, they are not enabled in BigQuery.

![image](https://user-images.githubusercontent.com/825732/198940726-e6877906-f19e-46da-a77f-74f3310f01d5.png)

![image](https://user-images.githubusercontent.com/825732/198940663-a5d512e5-a5f6-4406-bba4-3cf7e013b32b.png)

Our test also failed if we add `SELECT "abcd" LIKE ANY (SELECT "a%d")`.

```
failed to parse statements: failed to parse statement: INVALID_ARGUMENT: LIKE ANY is not supported [type.googleapis.com/zetasql.ErrorLocation='\x08\x01\x10\x14']
```

So for now we can't add such test, and `ast.SubqueryTypeLikeAny:` case will never be reached.

Does that sound good?

Reply:
ZetaSQL's Analyzer allows you to optionally specify which features to support. https://github.com/goccy/go-zetasqlite/blob/46f8b1037933aa6742331940b99e64cb531790c8/internal/analyzer.go#L49-L81

The following is the option in this case, which I think can be tested by enabling it.

https://github.com/goccy/go-zetasql/blob/50b559cf1db2019fd273c97a5338a3e4f1424952/enum.go#L666

Reply:
@goccy 

Just a quick report.

Test case:

```go
		{
			name:         "subquery expr with like any type",
			query:        `SELECT "abc" LIKE ANY ("ab%", "xy%")`,
			expectedRows: [][]interface{}{{true}},
		},
		{
			name:         "subquery expr with like some type",
			query:        `SELECT "abc" LIKE SOME ("ab%", "xy%")`,
			expectedRows: [][]interface{}{{true}},
		},
		{
			name:         "subquery expr with like all type",
			query:        `SELECT "abc" LIKE ALL ("a%", "ab%")`,
			expectedRows: [][]interface{}{{true}},
		},
```

Without `zetasql.FeatureV13LikeAnySomeAll` language feature:

```
--- FAIL: TestQuery (0.01s)
    --- FAIL: TestQuery/subquery_expr_with_like_any_type (0.00s)
        query_test.go:3281: failed to parse statements: failed to parse statement: INVALID_ARGUMENT: LIKE ANY is not supported [type.googleapis.com/zetasql.ErrorLocation='\x08\x01\x10\x13']
    --- FAIL: TestQuery/subquery_expr_with_like_some_type (0.00s)
        query_test.go:3281: failed to parse statements: failed to parse statement: INVALID_ARGUMENT: LIKE SOME is not supported [type.googleapis.com/zetasql.ErrorLocation='\x08\x01\x10\x13']
    --- FAIL: TestQuery/subquery_expr_with_like_all_type (0.00s)
        query_test.go:3281: failed to parse statements: failed to parse statement: INVALID_ARGUMENT: LIKE ALL is not supported [type.googleapis.com/zetasql.ErrorLocation='\x08\x01\x10\x13']
FAIL

```

With `zetasql.FeatureV13LikeAnySomeAll`, `INVALID_ARGUMENT: LIKE ANY is not supported ` error has gone but then another error:

```
--- FAIL: TestQuery (0.01s)
    --- FAIL: TestQuery/subquery_expr_with_like_any_type (0.00s)
        query_test.go:3281: failed to query SELECT zetasqlite_like_any("eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYWJjIn0=","eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYWIlIn0=","eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoieHklIn0=") AS `$col1#1` : no such function: zetasqlite_like_any
    --- FAIL: TestQuery/subquery_expr_with_like_some_type (0.00s)
        query_test.go:3281: failed to query SELECT zetasqlite_like_any("eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYWJjIn0=","eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYWIlIn0=","eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoieHklIn0=") AS `$col1#1` : no such function: zetasqlite_like_any
    --- FAIL: TestQuery/subquery_expr_with_like_all_type (0.00s)
        query_test.go:3281: failed to query SELECT zetasqlite_like_all("eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYWJjIn0=","eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYSUifQ==","eyJoZWFkZXIiOiJzdHJpbmciLCJib2R5IjoiYWIlIn0=") AS `$col1#1` : no such function: zetasqlite_like_all
FAIL
```

It seems that currently zetasqlite couldn't support `LIKE ANY` clauses and need to support them first.  
I can try to implement it but we should separate to another PR, I think ?

On the other hand, I'm not sure the priority of those supports because the real BigQuery doesn't support them neither.

Reply:
I see. Thank you for confirmation ! You don't need to add test cases for `ANY` and `ALL` .
I will support them later